test/functional/p2p_handshake.py

```diff
@@ -0,0 +1,49 @@
+#!/usr/bin/env python3
+# Copyright (c) 2024 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""
+Test P2P behaviour during the handshake phase (VERSION, VERACK messages).
+"""
+import random
+
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.messages import (
+    NODE_NETWORK,
+    NODE_NONE,
+    NODE_WITNESS,
+)
+from test_framework.p2p import P2PInterface
+
+
+DESIRABLE_SERVICE_FLAGS = NODE_NETWORK | NODE_WITNESS
+
+
+class P2PHandshakeTest(BitcoinTestFramework):
+    def set_test_params(self):
+        self.num_nodes = 1
+
+    def check_outbound_disconnect(self, node, p2p_idx, connection_type, services):
+        node.add_outbound_p2p_connection(
+            P2PInterface(), p2p_idx=p2p_idx, wait_for_disconnect=True,
+            connection_type=connection_type, services=services)
+
+    def run_test(self):
+        node = self.nodes[0]
+        self.log.info("Check that lacking desired service flags leads to disconnect")
+        for i, conn_type in enumerate(["outbound-full-relay", "block-relay-only", "addr-fetch"]):
+            services = random.choice([NODE_NONE, NODE_NETWORK, NODE_WITNESS])
```

Comment:
Why not enumerate all combinations of services and connection types here?

Reply:
ACK this nit. Any reason to not include all here?

Reply:
Benefits of enumerating: tests will be comprehensive and ensure that results can be consistently reproduced on every run, instead of being possibly flaky.
Costs: 9 instead of 3 iterations of this case.

Seems to weigh in favor of iteration - imo our tests should tend towards comprehensive and consistent.

Reply:
(ACKing nit).